clang/lib/Driver/Driver.cpp

```diff
@@ -3446,17 +3452,34 @@ class OffloadingActionBuilder final {
           if (!DeviceObjects.empty()) {
             // link and wrap the device binary, but do not perform the
             // backend compile.
+            // TODO: perform split here?
             auto *DeviceLinkAction =
-                C.MakeAction<LinkJobAction>(DeviceObjects, types::TY_SPIRV);
+                C.MakeAction<LinkJobAction>(DeviceObjects, types::TY_LLVM_BC);
+            auto *SPIRVTranslateAction = C.MakeAction<SPIRVTranslatorJobAction>(
+                DeviceLinkAction, types::TY_SPIRV);
             auto *DeviceWrappingAction = C.MakeAction<OffloadWrapperJobAction>(
-                DeviceLinkAction, types::TY_Object);
+                SPIRVTranslateAction, types::TY_Object);
             DA.add(*DeviceWrappingAction, **TC, /*BoundArch=*/nullptr,
                    Action::OFK_SYCL);
           }
           continue;
         }
         auto *DeviceLinkAction =
-            C.MakeAction<LinkJobAction>(LI, types::TY_SPIRV);
+            C.MakeAction<LinkJobAction>(LI, types::TY_LLVM_BC);
+        ActionList WrapperInputs;
+        Action *SPIRVInput = DeviceLinkAction;
+        if (!NoDeviceCodeSplit) {
+          auto *SplitAction = C.MakeAction<SYCLPostLinkJobAction>(
+              DeviceLinkAction, types::TY_TempIRfilelist);
+          auto *EntryGenAction = C.MakeAction<SYCLPostLinkJobAction>(
+              DeviceLinkAction, types::TY_TempEntriesfilelist);
+          SPIRVInput = SplitAction;
+          WrapperInputs.push_back(EntryGenAction);
+        }
+        auto *SPIRVTranslateAction = C.MakeAction<SPIRVTranslatorJobAction>(
+            SPIRVInput,
+            (NoDeviceCodeSplit) ? types::TY_SPIRV : types::TY_Tempfilelist);
```

Comment:
I'd suggest setting `TY_*` as a variable inside the preceding `if (!NoDeviceCodeSplit)`.

Reply:
Okay.